src/libtools/graphalgorithm.cc

```diff
@@ -54,6 +55,141 @@ bool singleNetwork(Graph g, GraphVisitor& gv) {
+          int degree = vertex_record.second.second;
+          if(degree>2){
+            return vertex_record.first;
+          }
+          remaining_unexplored.push_back(vertex_record.first);
+        }
+      }
+
+      // Search tips next
+      for(auto vertex : remaining_unexplored){
+        if(vertex_explored_[vertex].second==1){
+          return vertex;
+        }
+      }
+
+      // Finally if there are no tips or junctions left we will return a vertex
+      // of degree 2 if one exists
+      for(auto vertex : remaining_unexplored){
+        if(!vertex_explored_[vertex].first) return vertex;
+      }
+
+      throw runtime_error("You cannot get an unexplored vertex as they have all"
+          " been explored.");
+    }
+
+};
+
+ReducedGraph reduceGraph(Graph g){
+
+  unordered_map<int,pair<bool,int>> unexplored_vertices;
+  auto vertices = g.getVertices();
+  for(auto vertex : vertices ){
+    unexplored_vertices[vertex]=pair<bool,int>(false,g.getDegree(vertex));
+  }
+
+  ExplorationRecord exploration_record(unexplored_vertices);
+
+  vector<vector<int>> chains;
+  
+  while(exploration_record.unexploredVerticesExist()){
+    Graph_DF_Visitor gv;
+    int starting_vertex = exploration_record.getUnexploredVertex();
+    exploration_record.explore(starting_vertex);
+    gv.setStartingVertex(starting_vertex);
+    gv.initialize(g);
+
+    vector<int> chain{starting_vertex};
+    int old_vertex = starting_vertex;
+    bool new_chain = false;
+    while (!gv.queEmpty()) {
+      auto ed = gv.nextEdge(g);
+      
+      vector<int> unexplored_vertex = gv.getUnexploredVertex(ed);
+
+      if(new_chain){
+        if(unexplored_vertex.size()==0){
+          old_vertex = ed.getEndPoint1();       
+          chain.push_back(old_vertex);
+          new_chain = false;
+        }else{
+          old_vertex = ed.getOtherEndPoint(unexplored_vertex.at(0));
+          chain.push_back(old_vertex);
+          new_chain = false;
+        }
+      }
+      int new_vertex = ed.getOtherEndPoint(old_vertex);
+
+      if(unexplored_vertex.size()==0){
+        chain.push_back(new_vertex);
+        chains.push_back(chain);    
+        chain.clear();
+        new_chain = true; 
+      }else if(g.getDegree(new_vertex)==1){
+        chain.push_back(new_vertex);
+        chains.push_back(chain);
+        chain.clear();
+        exploration_record.explore(new_vertex);
+        new_chain = true; 
+      }else if(g.getDegree(new_vertex)>2){
+        chain.push_back(new_vertex);
+        chains.push_back(chain);
+        chain.clear();
+        exploration_record.explore(new_vertex);
+        new_chain = true; 
+      }else if(unexplored_vertex.size()==1){
+        chain.push_back(new_vertex); 
+        old_vertex = new_vertex;
+        exploration_record.explore(new_vertex);
+      }
+
+      gv.exec(g, ed);
+    }
+  }
+  vector<ReducedEdge> reduced_edges;
+  for(vector<int> chain : chains){
+    ReducedEdge reduced_ed(chain);
+    cout << reduced_ed << endl;
```

Comment:
why cout?

Reply:
It was left in from a debug session. I have removed it now.